src/code.lisp

```diff
@@ -7,21 +7,72 @@
            :reader read-query)))
 
 
-(defun postgres-query (query header)
-  (make 'postgres-query :query query :header header))
+(defclass postgres-simple-query (postgres-query)
+  ())
+
+
+(defclass postgres-prepared-query (postgres-query)
+  ((%base-parameters :initarg :base-parameters
+                     :reader read-base-parameters)
+   (%parameters :initarg :parameters
+                :reader read-parameters)
+   (%query-name :initarg :query-name
+                :reader read-query-name)))
+
+
+(defun postgres-simple-query (query header)
+  (make 'postgres-simple-query :query query :header header))
+
+
+(defun postgres-prepared-query (query base-parameters &optional query-name)
```

Comment:
Whatever happen to the header?

Reply:
Also, since query-name defaults to "" I would simply replace it with (query-name "") so check-type can be added in the method body.

Reply:
if we use `header` in this function, it wouldn't need to set `:columns` on the copy-from. 
I kept the header in `copy-from`, so we have a more familiar api. So we set the actual header there.

The 2 options (they all looks nice):

```lisp
(setf *ds*
  (prepare-query *query*  *params*
                 :columns '(id))) ; ?
               
(copy-from :postmodern *ds*)
```

```lisp
(setf *ds* (prepare-query *query*  *params*))
               
(copy-from :postmodern *ds* :columns '(id))
```

Reply:
I think we have a misunderstanding here. Header in the query is because this way query can be traversed in with-header scope with vellum:bind-row without actually materializing data-frame. I want to keep it this way. To solve this, copy-from should always accept just string or a list and construct either postgres-simple-query or postgres-prepared-query on it's own. It is simply put a higher level function. 

Perhaps way do do it is to accept parameters as KEY argument in COPY-FROM (remember that in CL you can have value, default value, and value to signal if value was provided, for instance (defun test (&optional (a "default" a-bound-p))).